lib/decompress/zstd_decompress_block.c

```diff
@@ -1112,44 +1202,210 @@ ZSTD_decompressSequences_body( ZSTD_DCtx* dctx,
+            __asm__(".p2align 6");
+#if !defined(__APPLE__)
+            __asm__("nop");
+            __asm__(".p2align 5");
+            __asm__("nop");
+#    if __GNUC__ >= 9 && __GNUC__ < 11
+            /* better for gcc-9 and gcc-10, worse for clang and gcc-8, gcc-11 */
+            __asm__(".p2align 3");
+#    else
+            __asm__(".p2align 4");
+#    endif
+#  endif
+#endif
+            for (; litPtr + sequence.litLength <= dctx->litBufferEnd; ) {
+                size_t const oneSeqSize = ZSTD_execSequence(op, oend, litPtr + sequence.litLength - WILDCOPY_OVERLENGTH, sequence, &litPtr, litEnd, prefixStart, vBase, dictEnd);
+#if defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION) && defined(FUZZING_ASSERT_VALID_SEQUENCE)
+                assert(!ZSTD_isError(oneSeqSize));
+                if (frame) ZSTD_assertValidSequence(dctx, op, oend, sequence, prefixStart, vBase);
+#endif
+                if (UNLIKELY(ZSTD_isError(oneSeqSize)))
+                    return oneSeqSize;
+                DEBUGLOG(6, "regenerated sequence size : %u", (U32)oneSeqSize);
+                op += oneSeqSize;
+                if (UNLIKELY(!--nbSeq))
+                    break;
+                BIT_reloadDStream(&(seqState.DStream));
+                sequence = ZSTD_decodeSequence(&seqState, isLongOffset);
+            }
+
+            if (nbSeq > 0) {
+                const BYTE* const oend_w = oend - WILDCOPY_OVERLENGTH;
+                const size_t leftoverLit = dctx->litBufferEnd - litPtr;
+                if (leftoverLit)
+                {
+                    RETURN_ERROR_IF(leftoverLit > (size_t)(oend - op), dstSize_tooSmall, "remaining lit must fit within dstBuffer");
+                    ZSTD_safecopy(op, oend_w, litPtr, leftoverLit, ZSTD_overlap_dst_before_src);
+                    sequence.litLength -= leftoverLit;
+                    op += leftoverLit;
+                }
+                litPtr = dctx->litExtraBuffer;
+                litEnd = dctx->litExtraBuffer + ZSTD_LITBUFFEREXTRASIZE;
+                dctx->splitLitBuffer = 0;
+                {
+                    size_t const oneSeqSize = ZSTD_execSequence(op, oend, oend_w, sequence, &litPtr, litEnd, prefixStart, vBase, dictEnd);
+#if defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION) && defined(FUZZING_ASSERT_VALID_SEQUENCE)
+                    assert(!ZSTD_isError(oneSeqSize));
+                    if (frame) ZSTD_assertValidSequence(dctx, op, oend, sequence, prefixStart, vBase);
+#endif
+                    if (UNLIKELY(ZSTD_isError(oneSeqSize)))
+                        return oneSeqSize;
+                    DEBUGLOG(6, "regenerated sequence size : %u", (U32)oneSeqSize);
+                    op += oneSeqSize;
+                    if (--nbSeq)
+                        BIT_reloadDStream(&(seqState.DStream));
+                }
+            }
+        }
+
+        if (nbSeq > 0) /* either lit buffer was from source, or there is remaining lit from extra buffer */
```

Comment:
So there are essentially 2 loops (explain this strategy in a comment).
OK, that's fine.
I would expect the second loop to deal with "the rest", which should be, hopefully, a small amount.
In which case, the complex alignment strategy below may not be necessary.